### PR TITLE
refactor: Pull out cargo-add MSRV code for reuse

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

#12078 added MSRV code in `cargo add`. Our assumption when writing it is that we'd need to generalize the code before reusing it in other places, like `cargo install`.  This PR focused purely on that refactor because I'm hopeful it will be useful for other work I'm doing.  Despite not having a user for this yet, I think the `cargo install` case is inevitable and I feel this does a bit to clean up MSRV related code by using a more specific type everywhere.

### How should we test and review this PR?

Each commit gradually progresses things along

### CODE REVIEW

File: src/cargo/util/semver_ext.rs [hunk 1]
```diff
@@ -108,6 +108,125 @@ impl From<VersionReq> for OptVersionReq {
...
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct VersionVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for VersionVisitor {
+            type Value = PartialVersion;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("semver version")
```
Comment (weihanglo) [inline src/cargo/util/semver_ext.rs:208]:
```suggestion
                formatter.write_str("SemVer version")
```

File: src/cargo/ops/registry/publish.rs [hunk 2]
```diff
@@ -424,7 +425,7 @@ fn transmit(
                 license_file: license_file.clone(),
                 badges: badges.clone(),
                 links: links.clone(),
-                rust_version: rust_version.clone(),
+                rust_version: rust_version,
```
Comment (weihanglo) [inline src/cargo/ops/registry/publish.rs:428]:
```suggestion
                rust_version,
```

File: src/cargo/util/semver_ext.rs [hunk 3]
```diff
@@ -108,6 +108,125 @@ impl From<VersionReq> for OptVersionReq {
     }
 }
 
+#[derive(PartialEq, Eq, Hash, Copy, Clone, Debug)]
+pub struct PartialVersion {
```
Comment (weihanglo) [inline src/cargo/util/semver_ext.rs:112]: Some prior art <https://github.com/oli-obk/cargo_metadata/pull/243>.

Do you feel like `BareVersion` is a better name? I am good with either.

**Review by weihanglo — COMMENTED** [review]
Looks good. The review comments aren't really important. Feel free to resolve them and I'll come back and merge.

File: src/cargo/util/semver_ext.rs (hunk 3 above)
Comment (epage) [reply src/cargo/util/semver_ext.rs:112]: For me, `Partial` better conveys than `Bare`.  If there is prior art for `Bare`, I'm fine with it.

Comment (weihanglo) [reply src/cargo/util/semver_ext.rs:112]: From [here](https://doc.rust-lang.org/nightly/cargo/reference/manifest.html#the-rust-version-field),

> The Rust version must be a bare version number with two or three components; 

Let's have a record and move on with `PartialVersion`. It's not worth too much efforts here :)

File: src/cargo/util/semver_ext.rs [hunk 4]
```diff
@@ -108,6 +108,125 @@ impl From<VersionReq> for OptVersionReq {
...
+
+            fn visit_str<E>(self, string: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                string.parse().map_err(serde::de::Error::custom)
+            }
+        }
+
+        deserializer.deserialize_str(VersionVisitor)
```
[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/util/semver_ext.rs]: In the tests, I'm not seeing a change in error message

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Thanks! 

@bors r+

**Pull request merged**